omnigibson/utils/usd_utils.py

```diff
@@ -1032,8 +1018,188 @@ def get_relative_jacobian(self, prim_path):
+        for view in cls._view_by_pattern.values():
+            view.clear()
+
+    @classmethod
+    def flush_control(cls):
+        for view in cls._view_by_pattern.values():
+            view.flush_control()
+
+    @classmethod
+    def initialize_view(cls):
+        cls._view_by_pattern = {}
+
+        # First, get all of the controllable objects in the scene (avoiding circular import)
+        from omnigibson.objects.controllable_object import ControllableObject
+
+        controllable_objects = [
+            obj for scene in og.sim.scenes for obj in scene.objects if isinstance(obj, ControllableObject)
+        ]
+
+        # Get their corresponding prim paths
+        expected_regular_prim_paths = {obj.articulation_root_path for obj in controllable_objects}
+        expected_dummy_prim_paths = {
+            obj._dummy.articulation_root_path
+            for obj in controllable_objects
+            if hasattr(obj, "_dummy") and obj._dummy is not None
+        }
+        expected_prim_paths = expected_regular_prim_paths | expected_dummy_prim_paths
+
+        # Group the prim paths by robot type
+        patterns = {cls._get_pattern_from_prim_path(prim_path) for prim_path in expected_prim_paths}
+
+        # Create the view for each robot type / fixedness combo
+        for pattern in patterns:
+            if pattern not in cls._view_by_pattern:
+                cls._view_by_pattern[pattern] = BatchControlViewAPIImpl(pattern)
+
+        # Initialize the views
+        for view in cls._view_by_pattern.values():
+            view.initialize_view()
+
+        # Assert that the views' prim paths are disjoint
+        all_prim_paths = []
+        for view in cls._view_by_pattern.values():
+            all_prim_paths.extend(view._idx.keys())
+        counts = collections.Counter(all_prim_paths)
+
+        missing = set(expected_prim_paths) - set(all_prim_paths)
+        assert len(missing) == 0, f"Prim paths {missing} are missing from the views!"
+
+        more_than_once = {prim_path: count for prim_path, count in counts.items() if count > 1}
+        assert len(more_than_once) == 0, f"Prim paths {more_than_once} are present in multiple views!"
+
+    @classmethod
+    def _get_pattern_from_prim_path(cls, prim_path):
+        """
+        Returns which of the regexes this prim path should be found in.
+
+        Note that since the prim path will be an articulation root path, this works for both fixed base and
+        non-fixed base robots. Fixed and non-fixed versions of the same robot will be mapped to different
+        patterns since they have different articulation root paths (object prim vs base link prim).
+        """
+        scene_id, robot_name = prim_path.split("/")[2:4]
+        assert scene_id.startswith("scene_"), f"Prim path 2nd component {prim_path} does not start with scene_"
+        components = robot_name.split("__")
+        assert len(components) == 3, f"Robot name {robot_name} does not match expected format"
```

Comment:
Can this be explicit -- what exactly do we expect the 3 components to be?

Reply:
Done.